### PR TITLE
Fix: Update thumbnail url and channel name for subscriptions when change detected

### LINKED ISSUE #1677: Channel icon in subscription list not updating with change
**Behavior of the program**
If you subscribe to a channel, it's icon appears in the subscription list, but, if the channel changes their icon then only the icon on the channel's page will change; icons in the subscription list don't get updated unless you manually unsubscribe and resubscribe.

**Expected behavior**
Checking every subscription for an icon change on every load (or even every couple of loads) would be inefficient from a network perspective, so instead I propose this:

The icon on the channel page changes as expected, so we can compare the icon loaded on the channel to the icon in the subscription list to see if the sub-list icon needs to be updated, the sub-list icon will only change when someone goes to the channel which is much nicer on the network then checking every channel at once.

**To Reproduce**
1. Subscribe to a channel.
2. Have that channel change it's icon.
3. See the channel's page display the new icon, but, not the subscription list.

**Environment Information**
 - OS: Arch Linux
 - FreeTube version: 0.15.1 Beta
 - Installation Method: FlatHub
 - Primary API Used: Local API [Happens with both]

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
---
Update thumbnail url for subscriptions when change detected
---

**Pull Request Type**
Please select what type of pull request this is:
- [x] Bugfix
- [ ] Feature Implementation

**Related issue**
Closes #1677

**Description**
When navigating to a channel, video or playlist page: it will check if the user is subscribed to the author & if they are subscribed then it will check if the thumbnail needs updating and will update the thumbnail accordingly

**Testing (for code that is not small enough to be easily understandable)**
I did the following for Invidious & local api:
I modified the thumbnail url to one of my subscriptions then:
- navigated to one of their videos (without navigating to channel)
- navigated to one of their playlists (without navigating to channel)
- navigated to their channel page.

**Desktop (please complete the following information):**
 - OS: Windows
 - OS Version: 10
 - FreeTube version: 0.15.0 Beta

**Additional context**

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
the function to update thumbnails is pretty repetitive so I'm wondering if there's a file I can move it to so it only appears once (I'd update the function to take in thumbnail url & channel Id)

**Comment by jasonhenriquez** [comment]
> the function to update thumbnails is pretty repetitive so I'm wondering if there's a file I can move it to so it only appears once (I'd update the function to take in thumbnail url & channel Id)

I use this implementation for global functions in the Accessibility Improvements PR: https://github.com/FreeTubeApp/FreeTube/blob/7391855a546d636fb05cfc21032f5ed4294fbf00/src/renderer/App.js#L477
Note that the function name begins with a `$` so as to explicitly denote that this function is global.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@ChunkyProgrammer can u resolve the merge conflicts. i wanted to test but im unable to because of it :(

**Comment by PrestonN** [comment]
> the function to update thumbnails is pretty repetitive so I'm wondering if there's a file I can move it to so it only appears once (I'd update the function to take in thumbnail url & channel Id)

I actually created the [Utils Store](https://github.com/FreeTubeApp/FreeTube/blob/development/src/renderer/store/modules/utils.js) for this exact reason.  The idea is that miscellaneous functions that are used across the app but don't have a good spot to go typically goes into utils.

I'm open to suggestions from you or @jasonhenriquez if this approach should be changed however, but that's how I've been handling it so far.

**Comment by ChunkyProgrammer** [comment]
Sounds good, I'll move the method to utils when I get the chance

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
wait a second isnt https://github.com/FreeTubeApp/FreeTube/issues/2204 the same as #1677?

**Comment by ChunkyProgrammer** [comment]
> wait a second isnt #2204 the same as #1677?

They are different. This PR should fix 1677. Another PR would have to fix 2204(lazyload/only load images when they are visible to user). This pr might fix some issues wirh 2204 though

File: src/renderer/views/Channel/Channel.js [hunk 1]
```diff
@@ -364,7 +367,9 @@ export default Vue.extend({
         } else {
           this.subCount = response.subCount
         }
-        this.thumbnailUrl = response.authorThumbnails[3].url.replace('https://yt3.ggpht.com', `${this.currentInvidiousInstance}/ggpht/`)
+        const thumbnail = response.authorThumbnails[3].url
+        this.thumbnailUrl = thumbnail.replace('https://yt3.ggpht.com', `${this.currentInvidiousInstance}/ggpht/`)
+        this.updateChannelThumbnail(thumbnail, this.channelName, this.id)
```
Comment (hockerschwan) [inline src/renderer/views/Channel/Channel.js:372]: This part worked with
```suggestion
        this.updateChannelThumbnail({
          channelThumbnailUrl: this.thumbnailUrl,
          channelName: this.channelName,
          channelId: this.id
        })
```

Comment (ChunkyProgrammer) [reply src/renderer/views/Channel/Channel.js:372]: Did not realize that. I'll make that change when I get the chance. Thanks.

Comment (ChunkyProgrammer) [reply src/renderer/views/Channel/Channel.js:372]: I think I did it like this so it saves the youtube thumbnail url and not the invidious one. Will update it to use the named parameters though

**Comment by absidue** [comment]
@ChunkyProgrammer please fix the merge conflicts.

**Review by absidue — APPROVED** [review]

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
Followed the testcases u provided and tried some hacky things to break it. LGTM!

**Comment by PikachuEXE** [comment]
How do you 
> I modified the thumbnail url to one of my subscriptions

**Comment by ChunkyProgrammer** [comment]
> How do you
> 
> > I modified the thumbnail url to one of my subscriptions

- Export Subscriptions as FreeTube subscriptions 
- Modify the thumbnail url for one of the exported subscriptions
- Remove subscription data (delete data settings, i think it was called?)
- Import the subscriptions file

**Review by PikachuEXE — COMMENTED** [review]
Tested locally
- navigated to one of their videos (without navigating to channel)
- navigated to one of their playlists (without navigating to channel)
- navigated to their channel page.

File: src/renderer/views/Playlist/Playlist.js [hunk 2]
```diff
@@ -80,6 +80,10 @@ export default Vue.extend({
           channelId: result.author ? result.author.channelID : '',
           infoSource: 'local'
         }
+        const channelThumbnailUrl = this.infoData.channelThumbnail.replace('=s48', '=s176')
```
Comment (PikachuEXE) [inline src/renderer/views/Playlist/Playlist.js:83]: I totally can't understand `replace('=s48', '=s176')` part
Can you add code comment to explain it

Same for other places with same code

File: src/renderer/store/modules/profiles.js [hunk 3]
```diff
@@ -89,6 +89,33 @@ const actions = {
     commit('setProfileList', profiles)
   },
 
+  async updateChannelThumbnail({ getters, dispatch }, { channelThumbnailUrl, channelName, channelId }) {
+    const profileList = getters.getProfileList
+    for (let i = 0; i < profileList.length; i++) {
```
Comment (PikachuEXE) [inline src/renderer/store/modules/profiles.js:94]: Is it possible to use `forEach` to replace for loop?
I only see `i` being used in `profileList[i]`

File: src/renderer/store/modules/profiles.js [hunk 4]
```diff
@@ -89,6 +89,33 @@ const actions = {
...
+      const currentProfile = JSON.parse(JSON.stringify(profileList[i]))
```
Comment (PikachuEXE) [inline src/renderer/store/modules/profiles.js:95]: This is a copy of "current profile"
`currentProfileCopy` might not be a good name but it's not the original object
```suggestion
      const currentProfileCopy = JSON.parse(JSON.stringify(profileList[i]))
```

File: src/renderer/store/modules/profiles.js [hunk 5]
```diff
@@ -89,6 +89,33 @@ const actions = {
...
+      const channelIndex = currentProfile.subscriptions.findIndex((channel) => {
+        return channel.id === channelId
+      })
+      if (channelIndex !== -1) {
+        console.log('channel found')
+        let updated = false
+        if (profileList[i].subscriptions[channelIndex].name !== channelName) {
```
Comment (PikachuEXE) [inline src/renderer/store/modules/profiles.js:102]: Why do you need to read from `profileList[i].subscriptions[channelIndex]` when you read from `currentProfile.subscriptions[channelIndex]`?

**Review by PikachuEXE — CHANGES_REQUESTED** [review]

**Comment by PikachuEXE** [comment]
@efb4f5ff-1298-471a-8973-3d47447115dc 
Feature review & code review are separate :)

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> @efb4f5ff-1298-471a-8973-3d47447115dc 
> Feature review & code review are separate :)

Yeah that's my bad for not noticing that 😬

File: src/renderer/store/modules/profiles.js (hunk 3 above)
Comment (ChunkyProgrammer) [reply src/renderer/store/modules/profiles.js:94]: it uses a break statement so i can't use a foreach

Comment (absidue) [reply src/renderer/store/modules/profiles.js:95]: you can use for-of in this case, if you want to:

```js
for (const profile of profileList) {
    // do stuff here with the profile variable
}
```

File: src/renderer/views/Playlist/Playlist.js [hunk 6]
```diff
@@ -135,6 +141,10 @@ export default Vue.extend({
           channelId: result.authorId,
           infoSource: 'invidious'
         }
+        const channelThumbnailUrl = result.authorThumbnails[2].url
+        const channelName = this.infoData.channelName
+        const channelId = this.infoData.channelId
+        this.updateSubscriptionDetails({ channelThumbnailUrl, channelName, channelId })
```
Comment (absidue) [inline src/renderer/views/Playlist/Playlist.js:147]: Please inline these variables as you did in https://github.com/FreeTubeApp/FreeTube/pull/1783/commits/ba537cc1b8572ce20467634bd5dcf7baf0b7eb67.

File: src/renderer/views/Watch/Watch.js [hunk 7]
```diff
@@ -276,6 +276,11 @@ export default Vue.extend({
             this.channelName = result.player_response.videoDetails.author
             this.channelThumbnail = result.player_response.embedPreview.thumbnailPreviewRenderer.videoDetails.embeddedPlayerOverlayVideoDetailsRenderer.channelThumbnail.thumbnails[0].url
           }
+          const channelName = this.channelName
+          const channelId = this.channelId
+          const channelThumbnailUrl = this.channelThumbnail
+          this.updateSubscriptionDetails({ channelThumbnailUrl, channelName, channelId })
```
Comment (absidue) [inline src/renderer/views/Watch/Watch.js:282]: Please inline these variables as well.

File: src/renderer/views/Watch/Watch.js [hunk 8]
```diff
@@ -561,7 +566,13 @@ export default Vue.extend({
           }
           this.channelId = result.authorId
           this.channelName = result.author
-          this.channelThumbnail = result.authorThumbnails[1] ? result.authorThumbnails[1].url.replace('https://yt3.ggpht.com', `${this.currentInvidiousInstance}/ggpht/`) : ''
+          const channelThumb = result.authorThumbnails[1]
+          this.channelThumbnail = channelThumb ? channelThumb.url.replace('https://yt3.ggpht.com', `${this.currentInvidiousInstance}/ggpht/`) : ''
+          const channelThumbnailUrl = channelThumb?.url
+          const channelName = this.channelName
+          const channelId = this.channelId
+          this.updateSubscriptionDetails({ channelThumbnailUrl, channelName, channelId })
```
Comment (absidue) [inline src/renderer/views/Watch/Watch.js:574]: Please inline these variables as well.

**Review by absidue — CHANGES_REQUESTED** [review]

File: src/renderer/views/Playlist/Playlist.js (hunk 2 above)
Comment (ChunkyProgrammer) [reply src/renderer/views/Playlist/Playlist.js]: the s is just the thumbnail size, i think

File: src/renderer/store/modules/profiles.js [hunk 9]
```diff
@@ -89,6 +89,33 @@ const actions = {
     commit('setProfileList', profiles)
   },
 
+  async updateSubscriptionDetails({ getters, dispatch }, { channelThumbnailUrl, channelName, channelId }) {
+    if (channelThumbnailUrl === null) {
```
Comment (PikachuEXE) [inline src/renderer/store/modules/profiles.js:93]: What do we handle `undefined` value?

File: src/renderer/store/modules/profiles.js [hunk 10]
```diff
@@ -89,6 +89,33 @@ const actions = {
...
+    const profileList = getters.getProfileList
+    for (const profile of profileList) {
+      const currentProfileCopy = JSON.parse(JSON.stringify(profile))
+      const channelIndex = currentProfileCopy.subscriptions.findIndex((channel) => {
+        return channel.id === channelId
+      })
+      if (channelIndex === 1) { continue }
+      let updated = false
+      const channel = currentProfileCopy.subscriptions[channelIndex]
+      if (channel.name !== channelName || (channel.thumbnail !== thumbnail && thumbnail !== null)) {
```
[the remark this answers is not in the document]
Comment (ChunkyProgrammer) [reply src/renderer/store/modules/profiles.js:106]: Potentially with this change #2283  or maybe if there's an api error/update

[the remark this answers is not in the document]
Comment (PikachuEXE) [reply src/renderer/store/modules/profiles.js:106]: I see `const thumbnail = channelThumbnailUrl.replace(/=s\d*/, '=s176')`
So I think this either will throw an error (when `channelThumbnailUrl` is not string) or it's returning a string
Am I missing anything? (Not sure)

File: src/renderer/store/modules/profiles.js (hunk 9 above)
Comment (ChunkyProgrammer) [reply src/renderer/store/modules/profiles.js]: fixed in latest commit (option chaining)

**Review by absidue — APPROVED** [review]

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
LGTM!

**Pull request merged**